### PR TITLE
Fix writes with 0 length

### LINKED ISSUE #220: Test fail on ITStorageTest.testUpdateBlob
After merging #211 travis build for `oraclejdk7` failed with an error in `ITStorageTest.testUpdate`:

```
testUpdateBlob(com.google.gcloud.storage.ITStorageTest)  Time elapsed: 10.421 sec  <<< ERROR!
com.google.gcloud.storage.StorageException: Backend Error
	at com.google.api.client.googleapis.json.GoogleJsonResponseException.from(GoogleJsonResponseException.java:145)
	at com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest.newExceptionOnError(AbstractGoogleJsonClientRequest.java:113)
	at com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest.newExceptionOnError(AbstractGoogleJsonClientRequest.java:40)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest$1.interceptResponse(AbstractGoogleClientRequest.java:321)
	at com.google.api.client.http.HttpRequest.execute(HttpRequest.java:1056)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:419)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:352)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.execute(AbstractGoogleClientRequest.java:469)
	at com.google.gcloud.spi.DefaultStorageRpc.patch(DefaultStorageRpc.java:231)
	at com.google.gcloud.storage.StorageImpl$11.call(StorageImpl.java:343)
	at com.google.gcloud.storage.StorageImpl$11.call(StorageImpl.java:340)
	at com.google.gcloud.RetryHelper.doRetry(RetryHelper.java:181)
	at com.google.gcloud.RetryHelper.runWithRetries(RetryHelper.java:247)
	at com.google.gcloud.RetryHelper.runWithRetries(RetryHelper.java:237)
	at com.google.gcloud.storage.StorageImpl.update(StorageImpl.java:340)
	at com.google.gcloud.storage.ITStorageTest.testUpdateBlob(ITStorageTest.java:150)
```

I never experienced this error locally and it seems not to occur always (builds for other java versions and next builds succeeded). The error message `Backend Error` makes me think that the problem resides outside of our library. I open this issue just to double check with you guys, can we consider this as a sporadic error?

### LINKED ISSUE #720: storage: "Failed to parse Content-Range header" when chunk-size upload is used
Here is the log:

Caused by: com.google.gcloud.storage.StorageException: 400 Bad Request
Failed to parse Content-Range header.
	at com.google.gcloud.spi.DefaultStorageRpc.translate(DefaultStorageRpc.java:94)
	at com.google.gcloud.spi.DefaultStorageRpc.write(DefaultStorageRpc.java:499)
	at com.google.gcloud.storage.BlobWriteChannel$1.run(BlobWriteChannel.java:49)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at com.google.gcloud.RetryHelper.doRetry(RetryHelper.java:181)
	at com.google.gcloud.RetryHelper.runWithRetries(RetryHelper.java:247)
	at com.google.gcloud.RetryHelper.runWithRetries(RetryHelper.java:237)
	at com.google.gcloud.storage.BlobWriteChannel.flushBuffer(BlobWriteChannel.java:46)
	at com.google.gcloud.BaseWriteChannel.close(BaseWriteChannel.java:147)

If I uploaded a file with size is exactly 2 x 1024 x 1024, and 1024 x 1024 as the chunk size, then I'll get above exception. However, if I changed the file size to 2 x 1024 x 1024 + 1, then the problem will go away.

Apparently, there is a bug in BlobWriteChannel->flushbuffer, if there is no data available, then there is no need to call options().rpc().write.

### PR DESCRIPTION
**Pull request by aozarov**
Fixes #720 and #723

### CODE REVIEW

File: gcloud-java-storage/src/test/java/com/google/gcloud/storage/it/ITStorageTest.java [hunk 1]
```diff
@@ -824,6 +824,29 @@ public void testReadAndWriteChannels() throws IOException {
...
+    String blobNamePrefix = "test-read-and-write-channels-blob-";
+    int[] blobSizes = {0, 700, 1024 * 256, 2 * 1024 * 1024, 4 * 1024 * 1024, 4 * 1024 * 1024 + 1};
+    Random rnd = new Random();
+    for (int blobSize : blobSizes) {
+      String blobName = blobNamePrefix + blobSize;
+      BlobInfo blob = BlobInfo.builder(BUCKET, blobName).build();
+      byte[] bytes = new byte[blobSize];
+      rnd.nextBytes(bytes);
+      try (WriteChannel writer = storage.writer(blob)) {
+        writer.write(ByteBuffer.wrap(BLOB_BYTE_CONTENT));
```
Comment (mziccard) [inline gcloud-java-storage/src/test/java/com/google/gcloud/storage/it/ITStorageTest.java]: Shouldn't it be `ByteBuffer.wrap(bytes)`?

File: gcloud-java-storage/src/test/java/com/google/gcloud/storage/it/ITStorageTest.java [hunk 2]
```diff
@@ -824,6 +824,29 @@ public void testReadAndWriteChannels() throws IOException {
...
+      }
+      ByteBuffer readBytes;
+      try (ReadChannel reader = storage.reader(blob.blobId())) {
+        readBytes = ByteBuffer.allocate(BLOB_BYTE_CONTENT.length);
```
Comment (mziccard) [inline gcloud-java-storage/src/test/java/com/google/gcloud/storage/it/ITStorageTest.java]: Also here (I might be missing something), should it be `ByteBuffer.allocate(blobSize)`?

File: gcloud-java-storage/src/test/java/com/google/gcloud/storage/it/ITStorageTest.java [hunk 3]
```diff
@@ -824,6 +824,29 @@ public void testReadAndWriteChannels() throws IOException {
...
+        reader.read(readBytes);
+      }
+      assertArrayEquals(BLOB_BYTE_CONTENT, readBytes.array());
```
Comment (mziccard) [inline gcloud-java-storage/src/test/java/com/google/gcloud/storage/it/ITStorageTest.java]: If previous comments are right this should also be `assertArrayEquals(bytes, readBytes.array());`

Comment (aozarov) [inline gcloud-java-storage/src/test/java/com/google/gcloud/storage/it/ITStorageTest.java]: All your comments are correct. My copy/paste of the testing code (that was done elsewhere) is completely wrong.
Will update the PR soon.

**Comment by mziccard** [comment]
Great job. Merging this!

**Pull request merged**